// @openai/openai/_shims/index.js
import { MultipartBody } from "./MultipartBody.js";
export const kind = 'web';
const _fetch = fetch;
export { _fetch as fetch };
const _Request = Request;
export { _Request as Request };
const _Response = Response;
export { _Response as Response };
const _Headers = Headers;
export { _Headers as Headers };
const _FormData = FormData;
export { _FormData as FormData };
const _File = File;
export { _File as File };
const _Blob = Blob;
export { _Blob as Blob };
export async function getMultipartRequestOptions(form, opts) {
  return {
    ...opts,
    body: new MultipartBody(form)
  };
}
export function getDefaultAgent(url) {
  return undefined;
}
export function fileFromPath() {
  throw new Error('The `fileFromPath` function is only supported in Node. See the README for more details: https://www.github.com/openai/openai-node#file-uploads');
}
export const isFsReadStream = (value)=>false;
const _ReadableStream = ReadableStream;
export { _ReadableStream as ReadableStream };
export const init = ()=>{};
